### PR TITLE
Demo_Ik

### PR DESCRIPTION
**Pull request by petrie911**
Not fully documented.

### CODE REVIEW

File: src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c [hunk 1]
```diff
@@ -9,27 +10,506 @@ void DemoIk_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+    SkelAnime_DrawFlexOpa(globalCtx, skelAnime->skeleton, skelAnime->jointTable, skelAnime->dListCount,
+                          DemoIk_Type2OverrideLimbDraw, DemoIk_Type2PostLimbDraw, this);
+    CLOSE_DISPS(gfxCtx, "../z_demo_ik_inFace.c", 341);
+}
+
+void DemoIk_Update(Actor* thisx, GlobalContext* globalCtx) {
+    static DemoIkActionFunc sActionFuncs[] = {
+        DemoIk_Type1Action0, DemoIk_Type1Action1, DemoIk_Type1Action2,
+        DemoIk_Type2Action0, DemoIk_Type2Action1, DemoIk_Type2Action2,
+    };
```
Comment (fig02) [inline src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c:484]: id keep these outside unless needed to match

File: src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c [hunk 2]
```diff
@@ -9,27 +10,506 @@ void DemoIk_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+        return;
+    }
+    sActionFuncs[this->actionMode](this, globalCtx);
+}
+
+void DemoIk_DrawNothing(DemoIk* this, GlobalContext* globalCtx) {
+}
+
+void DemoIk_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    static DemoIkDrawFunc sDrawFuncs[] = { DemoIk_DrawNothing, DemoIk_Type1Draw, DemoIk_Type2Draw };
```
Comment (fig02) [inline src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c:501]: same here

File: src/overlays/actors/ovl_Demo_Ik/z_demo_ik.h [hunk 3]
```diff
@@ -6,9 +6,17 @@
 
 struct DemoIk;
 
+typedef void (*DemoIkActionFunc) (struct DemoIk* this, GlobalContext* globalCtx);
+typedef void (*DemoIkDrawFunc) (struct DemoIk* this, GlobalContext* globalCtx);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Demo_Ik/z_demo_ik.h:10]:
```suggestion
typedef void (*DemoIkActionFunc)(struct DemoIk* this, GlobalContext* globalCtx);
typedef void (*DemoIkDrawFunc)(struct DemoIk* this, GlobalContext* globalCtx);
```

File: src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c [hunk 4]
```diff
@@ -9,27 +10,506 @@ void DemoIk_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+    this->drawMode = 0;
+    this->actor.shape.shadowAlpha = 0;
+}
+
+void func_8098394C(DemoIk* this, GlobalContext* globalCtx) {
+    DemoIk_EndMove(this);
+    DemoIk_MoveToStartPos(this, globalCtx, DemoIk_GetIndexFromParams(this->actor.params));
+    this->actionMode = 1;
+    this->drawMode = 1;
+    this->actor.shape.shadowAlpha = 0xFF;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c:217]:
```suggestion
    this->actor.shape.shadowAlpha = 255;
```

File: src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c [hunk 5]
```diff
@@ -9,27 +10,506 @@ void DemoIk_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+    this->skelAnime.curFrame = 0.0f;
+}
+
+void func_809839AC(DemoIk* this) {
+    this->actionMode = 2;
+    this->drawMode = 1;
+    this->actor.shape.shadowAlpha = 0xFF;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c:224]:
```suggestion
    this->actor.shape.shadowAlpha = 255;
```

File: src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c [hunk 6]
```diff
@@ -9,27 +10,506 @@ void DemoIk_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+
+void DemoIk_Type1PostLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, void* thisx) {
+    DemoIk* this = THIS;
+    GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;
+
+    OPEN_DISPS(gfxCtx, "../z_demo_ik_inArmer.c", 385);
+    if (limbIndex == 1) {
+        switch (this->actor.params) {
+            case 0:
+                gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(gfxCtx, "../z_demo_ik_inArmer.c", 390), 2);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Demo_Ik/z_demo_ik.c:287]: The last argument in all the gSPMatrix in this file need to be fixed to use the usual 3 gbi defines

**Review by fig02 — APPROVED** [review]
changes made

**Review by Roman971 — APPROVED** [review]

**Pull request merged**